compilation failure with fmt 10.1.1

with fmt 10.1.1 seastar now fails to compile with the error:

```
ninja: job failed: /usr/lib/ccache/bin/g++ -DFMT_SHARED -DSEASTAR_API_LEVEL=7 -DSEASTAR_BUILD_SHARED_LIBS -DSEASTAR_DEFERRED_ACTION_REQUIRE_NOEXCEPT -DSEASTAR_HAS_MEMBARRIER -DSEASTAR_HAVE_HWLOC -DSEASTAR_HAVE_NUMA -DSEASTAR_HAVE_URING -DSEASTAR_SCHEDULING_GROUPS_COUNT=16 -DSEASTAR_TESTING_MAIN -DSEASTAR_TESTING_WITH_NETWORKING=1 -I/home/dunk/gits/aports/testing/seastar/src/seastar-1ddb6fe061aea521aeeafe3ad6ab926b803ab51d/tests/unit -I/home/dunk/gits/aports/testing/seastar/src/seastar-1ddb6fe061aea521aeeafe3ad6ab926b803ab51d/src -I/home/dunk/gits/aports/testing/seastar/src/seastar-1ddb6fe061aea521aeeafe3ad6ab926b803ab51d/include -I/home/dunk/gits/aports/testing/seastar/src/seastar-1ddb6fe061aea521aeeafe3ad6ab926b803ab51d/build/gen/include -Os -fstack-clash-protection -Wformat -Werror=format-security -D_GLIBCXX_ASSERTIONS=1 -fno-plt -DSEASTAR_DEFAULT_ALLOCATOR -Os -DNDEBUG -std=gnu++20 -UNDEBUG -Wall -Wimplicit-fallthrough -Wno-array-bounds -Wno-error=deprecated-declarations -gz -U_FORTIFY_SOURCE -Wno-maybe-uninitialized -DSEASTAR_SSTRING -Wno-error=unused-result -MD -MT tests/unit/CMakeFiles/test_unit_smp.dir/smp_test.cc.o -MF tests/unit/CMakeFiles/test_unit_smp.dir/smp_test.cc.o.d -o tests/unit/CMakeFiles/test_unit_smp.dir/smp_test.cc.o -c /home/dunk/gits/aports/testing/seastar/src/seastar-1ddb6fe061aea521aeeafe3ad6ab926b803ab51d/tests/unit/smp_test.cc
In file included from /home/dunk/gits/aports/testing/seastar/src/seastar-1ddb6fe061aea521aeeafe3ad6ab926b803ab51d/include/seastar/core/scheduling.hh:29,
                 from /home/dunk/gits/aports/testing/seastar/src/seastar-1ddb6fe061aea521aeeafe3ad6ab926b803ab51d/include/seastar/core/task.hh:24,
                 from /home/dunk/gits/aports/testing/seastar/src/seastar-1ddb6fe061aea521aeeafe3ad6ab926b803ab51d/include/seastar/core/future.hh:36,
                 from /home/dunk/gits/aports/testing/seastar/src/seastar-1ddb6fe061aea521aeeafe3ad6ab926b803ab51d/include/seastar/core/timer.hh:24,
                 from /home/dunk/gits/aports/testing/seastar/src/seastar-1ddb6fe061aea521aeeafe3ad6ab926b803ab51d/include/seastar/core/condition-variable.hh:31,
                 from /home/dunk/gits/aports/testing/seastar/src/seastar-1ddb6fe061aea521aeeafe3ad6ab926b803ab51d/include/seastar/core/reactor.hh:28,
                 from /home/dunk/gits/aports/testing/seastar/src/seastar-1ddb6fe061aea521aeeafe3ad6ab926b803ab51d/tests/unit/smp_test.cc:22:
/home/dunk/gits/aports/testing/seastar/src/seastar-1ddb6fe061aea521aeeafe3ad6ab926b803ab51d/include/seastar/core/sstring.hh: In instantiation of 'auto fmt::v10::formatter<seastar::basic_sstring<char_type, Size, max_size, NulTerminate> >::format(const seastar::basic_sstring<char_type, Size, max_size, NulTerminate>&, FormatContext&) const [with FormatContext = fmt::v10::basic_format_context<fmt::v10::appender, char>; char_type = char; Size = unsigned int; Size max_size = 15; bool NulTerminate = true]':
/usr/include/fmt/core.h:1097:70:   required by substitution of 'template<class Context, class T> constexpr decltype ((typename Context::formatter_type<T>().format(declval<const T&>(), declval<Context&>()), true)) fmt::v10::detail::has_const_formatter_impl(T*) [with Context = fmt::v10::basic_format_context<fmt::v10::appender, char>; T = seastar::basic_sstring<char, unsigned int, 15>]'
/usr/include/fmt/core.h:1108:43:   required from 'constexpr bool fmt::v10::detail::has_const_formatter() [with T = seastar::basic_sstring<char, unsigned int, 15>; Context = fmt::v10::basic_format_context<fmt::v10::appender, char>]'
/usr/include/fmt/core.h:1443:69:   required from 'struct fmt::v10::detail::arg_mapper<fmt::v10::basic_format_context<fmt::v10::appender, char> >::formattable<const seastar::basic_sstring<char, unsigned int, 15>, seastar::basic_sstring<char, unsigned int, 15> >'
/usr/include/fmt/core.h:1451:25:   required by substitution of 'template<class T, typename std::enable_if<(! fmt::v10::detail::arg_mapper<fmt::v10::basic_format_context<fmt::v10::appender, char> >::formattable<T, typename std::remove_const<_Tp>::type>::value), int>::type <anonymous> > constexpr fmt::v10::detail::unformattable fmt::v10::detail::arg_mapper<fmt::v10::basic_format_context<fmt::v10::appender, char> >::do_map(T&) [with T = const seastar::basic_sstring<char, unsigned int, 15>; typename std::enable_if<(! fmt::v10::detail::arg_mapper<fmt::v10::basic_format_context<fmt::v10::appender, char> >::formattable<T, typename std::remove_const<_Tp>::type>::value), int>::type <anonymous> = <missing>]'
/usr/include/fmt/core.h:1462:69:   required by substitution of 'template<class T, class U, typename std::enable_if<((((((std::is_class<U>::value || std::is_enum<_Tp2>::value) || std::is_union<U>::value) && (! fmt::v10::detail::is_string<U>::value)) && (! fmt::v10::is_char<U>::value)) && (! fmt::v10::detail::is_named_arg<T>::value)) && (! std::is_arithmetic<typename fmt::v10::detail::format_as_result<U>::type>::value)), int>::type <anonymous> > constexpr decltype (((fmt::v10::detail::arg_mapper<fmt::v10::basic_format_context<fmt::v10::appender, char> >*)this)->fmt::v10::detail::arg_mapper<fmt::v10::basic_format_context<fmt::v10::appender, char> >::do_map(val)) fmt::v10::detail::arg_mapper<fmt::v10::basic_format_context<fmt::v10::appender, char> >::map(T&) [with T = const seastar::basic_sstring<char, unsigned int, 15>; U = seastar::basic_sstring<char, unsigned int, 15>; typename std::enable_if<((((((std::is_class<U>::value || std::is_enum<_Tp2>::value) || std::is_union<U>::value) && (! fmt::v10::detail::is_string<U>::value)) && (! fmt::v10::is_char<U>::value)) && (! fmt::v10::detail::is_named_arg<T>::value)) && (! std::is_arithmetic<typename fmt::v10::detail::format_as_result<U>::type>::value)), int>::type <anonymous> = 0]'
/usr/include/fmt/core.h:1478:53:   required by substitution of 'template<class T, class Context> using fmt::v10::detail::mapped_type_constant = fmt::v10::detail::type_constant<decltype (fmt::v10::detail::arg_mapper<Context>().map(declval<const T&>())), typename Context::char_type> [with T = seastar::basic_sstring<char, unsigned int, 15>; Context = fmt::v10::basic_format_context<fmt::v10::appender, char>]'
/usr/include/fmt/core.h:2609:66:   required from 'constexpr fmt::v10::detail::format_string_checker<Char, Args>::format_string_checker(fmt::v10::basic_string_view<Char>) [with Char = char; Args = {char [5], seastar::basic_sstring<char, unsigned int, 15, true>}]'
/home/dunk/gits/aports/testing/seastar/src/seastar-1ddb6fe061aea521aeeafe3ad6ab926b803ab51d/tests/unit/smp_test.cc:66:19:   required from here
/home/dunk/gits/aports/testing/seastar/src/seastar-1ddb6fe061aea521aeeafe3ad6ab926b803ab51d/tests/unit/smp_test.cc:66:19:   in 'constexpr' expansion of 'fmt::v10::basic_format_string<char, const char (&)[5], const seastar::basic_sstring<char, unsigned int, 15, true>&>("{}: {}\012")'
/home/dunk/gits/aports/testing/seastar/src/seastar-1ddb6fe061aea521aeeafe3ad6ab926b803ab51d/include/seastar/core/sstring.hh:888:68: error: no matching function for call to 'fmt::v10::formatter<seastar::basic_sstring<char, unsigned int, 15>, char, void>::format(const seastar::basic_sstring<char, unsigned int, 15>&, fmt::v10::basic_format_context<fmt::v10::appender, char>&) const'
  888 |         return formatter<std::basic_string_view<char_type>>::format(s, ctx);
      |                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~
In file included from /usr/include/fmt/format.h:49,
                 from /usr/include/fmt/ostream.h:20,
                 from /home/dunk/gits/aports/testing/seastar/src/seastar-1ddb6fe061aea521aeeafe3ad6ab926b803ab51d/include/seastar/core/sstring.hh:41:
/usr/include/fmt/core.h:2712:22: note: candidate: 'template<class FormatContext> constexpr decltype (ctx.out()) fmt::v10::formatter<T, Char, typename std::enable_if<(fmt::v10::detail::type_constant<T, Char>::value != fmt::v10::detail::type::custom_type), void>::type>::format(const T&, FormatContext&) const [with T = fmt::v10::basic_string_view<char>; Char = char]'
 2712 |   FMT_CONSTEXPR auto format(const T& val, FormatContext& ctx) const
      |                      ^~~~~~
/usr/include/fmt/core.h:2712:22: note:   template argument deduction/substitution failed:
/home/dunk/gits/aports/testing/seastar/src/seastar-1ddb6fe061aea521aeeafe3ad6ab926b803ab51d/include/seastar/core/sstring.hh:888:68: note:   cannot convert 's' (type 'const seastar::basic_sstring<char, unsigned int, 15>') to type 'const fmt::v10::basic_string_view<char>&'
  888 |         return formatter<std::basic_string_view<char_type>>::format(s, ctx);
      |                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~
/usr/include/fmt/core.h: In instantiation of 'constexpr fmt::v10::detail::value<Context> fmt::v10::detail::make_arg(T&) [with bool PACKED = true; Context = fmt::v10::basic_format_context<fmt::v10::appender, char>; T = const seastar::basic_sstring<char, unsigned int, 15>; typename std::enable_if<PACKED, int>::type <anonymous> = 0]':
/usr/include/fmt/core.h:1808:51:   required from 'constexpr fmt::v10::format_arg_store<Context, Args>::format_arg_store(T& ...) [with T = {const char [5], const seastar::basic_sstring<char, unsigned int, 15, true>}; Context = fmt::v10::basic_format_context<fmt::v10::appender, char>; Args = {char [5], seastar::basic_sstring<char, unsigned int, 15, true>}]'
/usr/include/fmt/core.h:1826:18:   required from 'constexpr fmt::v10::format_arg_store<Context, typename std::remove_cv<typename std::remove_reference<_Args>::type>::type ...> fmt::v10::make_format_args(T& ...) [with Context = basic_format_context<appender, char>; T = {const char [5], const seastar::basic_sstring<char, unsigned int, 15, true>}]'
/usr/include/fmt/core.h:2875:44:   required from 'void fmt::v10::print(format_string<T ...>, T&& ...) [with T = {const char (&)[5], const seastar::basic_sstring<char, unsigned int, 15, true>&}; format_string<T ...> = basic_format_string<char, const char (&)[5], const seastar::basic_sstring<char, unsigned int, 15, true>&>]'
/home/dunk/gits/aports/testing/seastar/src/seastar-1ddb6fe061aea521aeeafe3ad6ab926b803ab51d/tests/unit/smp_test.cc:66:19:   required from here
/usr/include/fmt/core.h:1576:63: error: 'fmt::v10::detail::type_is_unformattable_for<const seastar::basic_sstring<char, unsigned int, 15>, char> _' has incomplete type
 1576 |     type_is_unformattable_for<T, typename Context::char_type> _;
      |                                                               ^
/usr/include/fmt/core.h:1580:7: error: static assertion failed: Cannot format an argument. To make type T formattable provide a formatter<T> specialization: https://fmt.dev/latest/api.html#udt
 1580 |       formattable,
      |       ^~~~~~~~~~~
/usr/include/fmt/core.h:1580:7: note: 'formattable' evaluates to false
ninja: subcommands failed
```

I think some implicit conversion have been removed by fmt now

--- a/include/seastar/core/sstring.hh
+++ b/include/seastar/core/sstring.hh
@@ -882,10 +882,14 @@ std::ostream& operator<<(std::ostream& os, const std::unordered_map<Key, T, Hash
 
 SEASTAR_MODULE_EXPORT
 template <typename char_type, typename Size, Size max_size, bool NulTerminate>
-struct fmt::formatter<seastar::basic_sstring<char_type, Size, max_size, NulTerminate>> : fmt::formatter<std::basic_string_view<char_type>> {
+struct fmt::formatter<seastar::basic_sstring<char_type, Size, max_size, NulTerminate>>
+    : private fmt::formatter<std::basic_string_view<char_type>> {
+    using format_as_t = std::basic_string_view<char_type>;
+    using base = fmt::formatter<format_as_t>;
+    using base::parse;
     template <typename FormatContext>
     auto format(const ::seastar::basic_sstring<char_type, Size, max_size, NulTerminate>& s, FormatContext& ctx) const {
-        return formatter<std::basic_string_view<char_type>>::format(s, ctx);
+        return base::format(format_as_t{s}, ctx);
     }
 };
 

